File: packages/streamr-libstreamrproxyclient/test/integration/StreamrProxyClientTest.cpp
#include "streamrproxyclient.h"
#include <iostream>
#include <string>
#include <gtest/gtest.h>
import streamr.logger.SLogger;

using streamr::logger::SLogger;

class StreamrProxyClientTest : public ::testing::Test {
protected:
    static constexpr const char* invalidEthereumAddress =
        "INVALID_ETHEREUM_ADDRESS";
    static constexpr const char* goodEthereumAddress =
        "0x123456789012345678901234567890123456789a";
    static constexpr const char* validEthereumAddress =
        "0x1234567890123456789012345678901234567890";
    static constexpr const char* validEthereumAddress2 =
        "0x1234567890123456789012345678901234567892";
    static constexpr const char* validEthereumAddress3 =
        "0x1234567890123456789012345678901234567893";
    static constexpr const char* invalidStreamPartId = "INVALID_STREAM_PART_ID";
    static constexpr const char* validStreamPartId =
        "0xa000000000000000000000000000000000000000#01";

    static constexpr const char* invalidProxyUrl = "poiejrg039utg240";
    static constexpr const char* validProxyUrl = "ws://valid.com";

    static constexpr const char* nonExistentProxyUrl0 = "ws://localhost:0";
    static constexpr const char* nonExistentProxyUrl1 = "ws://localhost:1";
    static constexpr const char* nonExistentProxyUrl2 = "ws://localhost:2";

    static constexpr uint64_t invalidClientHandle = 0;

public:
    ~StreamrProxyClientTest() override { streamrCleanupLibrary(); }
};

TEST_F(StreamrProxyClientTest, CanCallApi) {
    std::cout << "RPC returned: " << testRpc() << "\n";
}

TEST_F(StreamrProxyClientTest, CanCreateAndDeleteProxyClient) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = validEthereumAddress;
    const char* streamPartId = validStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    EXPECT_EQ(result->numErrors, 0);
    EXPECT_NE(clientHandle, 0);
    streamrResultDelete(result);

    const StreamrResult* result2 = nullptr;
    proxyClientDelete(&result2, clientHandle);
    SLogger::info("proxyClientDelete called");
    EXPECT_EQ(result2->numErrors, 0);
    streamrResultDelete(result2);

    SLogger::info("test finished");
}

// Regression test for use-after-cleanup: the fixture destructor calls
// streamrCleanupLibrary() after every test, so any test that runs
// after another one in the same process starts with the library torn
// down. API entry points must re-initialize the library on demand
// instead of dereferencing the destroyed instance (this crashed with
// SIGSEGV before the fix; CI masked it by running every test in its
// own process).
TEST_F(StreamrProxyClientTest, ApiIsUsableAfterCleanupLibrary) {
    // Implicit re-initialization: call the API right after a cleanup.
    streamrCleanupLibrary();

    const StreamrResult* result = nullptr;
    uint64_t clientHandle =
        proxyClientNew(&result, validEthereumAddress, validStreamPartId);
    EXPECT_EQ(result->numErrors, 0);
    EXPECT_NE(clientHandle, 0);
    streamrResultDelete(result);

    const StreamrResult* result2 = nullptr;
    proxyClientDelete(&result2, clientHandle);
    EXPECT_EQ(result2->numErrors, 0);
    streamrResultDelete(result2);

    // Explicit re-initialization (the path documented in
    // streamrcommon.h): cleanup, init, use.
    streamrCleanupLibrary();
    streamrInitLibrary();

    const StreamrResult* result3 = nullptr;
    uint64_t clientHandle2 =
        proxyClientNew(&result3, validEthereumAddress, validStreamPartId);
    EXPECT_EQ(result3->numErrors, 0);
    EXPECT_NE(clientHandle2, 0);
    streamrResultDelete(result3);

    const StreamrResult* result4 = nullptr;
    proxyClientDelete(&result4, clientHandle2);
    EXPECT_EQ(result4->numErrors, 0);
    streamrResultDelete(result4);
}

TEST_F(StreamrProxyClientTest, InvalidEthereumAddress) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = invalidEthereumAddress;
    const char* streamPartId = validStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    SLogger::info("numErrors: " + std::to_string(result->numErrors));
    SLogger::info("errors: " + std::string(result->errors->message));
    SLogger::info("errors: " + std::string(result->errors->code));

    EXPECT_EQ(result->numErrors, 1);
    EXPECT_STREQ(result->errors->code, ERROR_INVALID_ETHEREUM_ADDRESS);
    streamrResultDelete(result);
}

TEST_F(StreamrProxyClientTest, InvalidStreamPartId) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = validEthereumAddress;
    const char* streamPartId = invalidStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    SLogger::info("numErrors: " + std::to_string(result->numErrors));
    SLogger::info("errors: " + std::string(result->errors->message));
    SLogger::info("errors: " + std::string(result->errors->code));

    EXPECT_EQ(result->numErrors, 1);
    EXPECT_STREQ(result->errors->code, ERROR_INVALID_STREAM_PART_ID);
    streamrResultDelete(result);
}

TEST_F(StreamrProxyClientTest, ProxyClientNotFound) {
    const StreamrResult* result = nullptr;

    StreamrPeer proxy{
        .websocketUrl = invalidProxyUrl,
        .ethereumAddress = validEthereumAddress};

    proxyClientConnect(&result, invalidClientHandle, &proxy, 1);

    EXPECT_EQ(result->numErrors, 1);
    EXPECT_STREQ(result->errors->code, ERROR_PROXY_CLIENT_NOT_FOUND);
    streamrResultDelete(result);
}

TEST_F(StreamrProxyClientTest, NoProxiesDefined) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = validEthereumAddress;
    const char* streamPartId = validStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    proxyClientConnect(&result, clientHandle, nullptr, 0);

    EXPECT_EQ(result->numErrors, 1);
    EXPECT_STREQ(result->errors->code, ERROR_NO_PROXIES_DEFINED);
    streamrResultDelete(result);
}

TEST_F(StreamrProxyClientTest, InvalidProxyUrl) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = validEthereumAddress;
    const char* streamPartId = validStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    streamrResultDelete(result);

    const StreamrResult* result2 = nullptr;
    StreamrPeer proxy{
        .websocketUrl = invalidProxyUrl,
        .ethereumAddress = validEthereumAddress};
    proxyClientConnect(&result2, clientHandle, &proxy, 1);

    EXPECT_EQ(result2->numErrors, 1);
    EXPECT_STREQ(result2->errors->code, ERROR_INVALID_PROXY_URL);
    streamrResultDelete(result2);

    const StreamrResult* result3 = nullptr;
    proxyClientDelete(&result3, clientHandle);
    streamrResultDelete(result3);
}

TEST_F(StreamrProxyClientTest, InvalidProxyEthereumAddress) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = validEthereumAddress;
    const char* streamPartId = validStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    const StreamrResult* result2 = nullptr;
    StreamrPeer proxy{
        .websocketUrl = validProxyUrl,
        .ethereumAddress = invalidEthereumAddress};
    proxyClientConnect(&result2, clientHandle, &proxy, 1);

    EXPECT_EQ(result2->numErrors, 1);
    EXPECT_STREQ(result2->errors->code, ERROR_INVALID_ETHEREUM_ADDRESS);
    streamrResultDelete(result2);

    const StreamrResult* result3 = nullptr;
    proxyClientDelete(&result3, clientHandle);
    streamrResultDelete(result3);
}

TEST_F(StreamrProxyClientTest, ProxyConnectionFailed) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = validEthereumAddress;
    const char* streamPartId = validStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    streamrResultDelete(result);
    const StreamrResult* result2 = nullptr;
    StreamrPeer proxy{
        .websocketUrl = nonExistentProxyUrl0,
        .ethereumAddress = validEthereumAddress};
    proxyClientConnect(&result2, clientHandle, &proxy, 1);

    SLogger::info("numErrors: " + std::to_string(result2->numErrors));
    SLogger::info("errors: " + std::string(result2->errors->message));
    SLogger::info("errors: " + std::string(result2->errors->code));

    EXPECT_EQ(result2->numErrors, 1);
    EXPECT_STREQ(result2->errors->code, ERROR_PROXY_CONNECTION_FAILED);
    streamrResultDelete(result2);

    const StreamrResult* result3 = nullptr;
    proxyClientDelete(&result3, clientHandle);
    streamrResultDelete(result3);
}

TEST_F(StreamrProxyClientTest, ThreeProxyConnectionsFailed) {
    const StreamrResult* result = nullptr;

    const char* ownEthereumAddress = goodEthereumAddress;
    const char* streamPartId = validStreamPartId;

    uint64_t clientHandle =
        proxyClientNew(&result, ownEthereumAddress, streamPartId);

    streamrResultDelete(result);

    // NOLINTNEXTLINE
    StreamrPeer proxies[] = {
        {.websocketUrl = nonExistentProxyUrl0,
         .ethereumAddress = validEthereumAddress},
        {.websocketUrl = nonExistentProxyUrl1,
         .ethereumAddress = validEthereumAddress2},
        {.websocketUrl = nonExistentProxyUrl2,
         .ethereumAddress = validEthereumAddress3}};

    const StreamrResult* result2 = nullptr;
    auto numConnections =
        proxyClientConnect(&result2, clientHandle, proxies, 3);

    EXPECT_EQ(numConnections, 0);
    EXPECT_EQ(result2->numErrors, 3);
    EXPECT_EQ(
        std::string(result2->errors[0].proxy->websocketUrl),
        std::string(proxies[0].websocketUrl));
    EXPECT_EQ(
        std::string(result2->errors[1].proxy->websocketUrl),
        std::string(proxies[1].websocketUrl));
    EXPECT_EQ(
        std::string(result2->errors[2].proxy->websocketUrl),
        std::string(proxies[2].websocketUrl));

    for (uint64_t i = 0; i < result2->numErrors; ++i) {
        SLogger::info("errors: " + std::string(result2->errors[i].message));
        EXPECT_STREQ(result2->errors[i].code, ERROR_PROXY_CONNECTION_FAILED);
    }
    const StreamrResult* result3 = nullptr;
    proxyClientDelete(&result3, clientHandle);
    streamrResultDelete(result3);
}

// The pre-3.0 proxy-prefixed names must stay source- and ABI-compatible
// for one release (phase D3a): this test compiles against the deprecated
// aliases and calls the deprecated symbols on purpose.
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
TEST(StreamrProxyClientDeprecatedAliasTest, Pre30NamesStillWork) {
    proxyClientInitLibrary();
    const ProxyResult* result = nullptr;
    Proxy proxy{
        .websocketUrl = "ws://127.0.0.1:44211",
        .ethereumAddress = "0x1234567890123456789012345678901234567890"};
    (void)proxy;
    const uint64_t handle = proxyClientNew(
        &result,
        "0xa5374e3c19f15e1847881979dd0c6c9ffe846bd5",
        "0xd2078dc2d780029473a39ce873fc182587be69db/low-level-client#0");
    ASSERT_NE(result, nullptr);
    EXPECT_EQ(result->numErrors, 0);
    proxyClientResultDelete(result);
    result = nullptr;
    proxyClientDelete(&result, handle);
    ASSERT_NE(result, nullptr);
    // The Error alias and its pre-3.0 `proxy` member spelling stay usable.
    if (result->numErrors > 0) {
        const Error* firstError = &result->errors[0];
        (void)firstError->proxy;
    }
    proxyClientResultDelete(result);
}
#pragma clang diagnostic pop

File: packages/streamr-libstreamrproxyclient/wrappers/go/streamrproxyclient_test.go
package streamrproxyclient

import (
	"fmt"
	"testing"
)

const (
	invalidEthereumAddress     = "INVALID_ETHEREUM_ADDRESS"
	goodEthereumAddress        = "0x123456789012345678901234567890123456789a"
	validEthereumAddress       = "0x1234567890123456789012345678901234567890"
	validEthereumAddress2      = "0x1234567890123456789012345678901234567892"
	validEthereumAddress3      = "0x1234567890123456789012345678901234567893"
	invalidStreamPartId        = "INVALID_STREAM_PART_ID"
	validStreamPartId          = "0xa000000000000000000000000000000000000000#01"
	ownEthereumAddress         = "0x1234567890123456789012345678901234567890"
	streamPartId               = "0xd2078dc2d780029473a39ce873fc182587be69db/low-level-client#0"
	invalidProxyUrl            = "poiejrg039utg240"
	validProxyUrl              = "ws://valid.com"
	nonExistentProxyUrl0       = "ws://localhost:0"
	nonExistentProxyUrl1       = "ws://localhost:1"
	nonExistentProxyUrl2       = "ws://localhost:2"
	invalidClientHandle uint64 = 0
)
func TestLibStreamrProxyClientCreation(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    if lib == nil {
        t.Fatalf("Failed to create LibStreamrProxyClient")
    }
    defer lib.Close()
    client, err := NewProxyClient(ownEthereumAddress, streamPartId)
    if err != nil {
        t.Fatalf("Failed to create ProxyClient: %v", err)
    }
    defer client.Close()
}

func TestInvalidEthereumAddress(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    defer lib.Close()
    client, err := NewProxyClient(invalidEthereumAddress, validStreamPartId)
    if err == nil {
        t.Fatalf("Expected error for invalid Ethereum address")
    } else if err.code != ERROR_INVALID_ETHEREUM_ADDRESS {
        t.Fatalf("Expected error code %s, got %s", ERROR_INVALID_ETHEREUM_ADDRESS, err.code)
    }
    if client != nil {
        defer client.Close()
        t.Fatalf("Expected nil client for invalid Ethereum address")
    }
}

func TestInvalidStreamPartId(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    defer lib.Close()
    client, err := NewProxyClient(validEthereumAddress, invalidStreamPartId)
    if err == nil {
        t.Fatalf("Expected error for invalid stream part id")
    } else if err.code != ERROR_INVALID_STREAM_PART_ID {
        t.Fatalf("Expected error code %s, got %s", ERROR_INVALID_STREAM_PART_ID, err.code)
    }
    if client != nil {
        defer client.Close()
        t.Fatalf("Expected nil client for invalid stream part id")
    }
}

func TestNoProxiesDefined(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    defer lib.Close()
    client, err := NewProxyClient(validEthereumAddress, validStreamPartId)
    if err != nil {
        t.Fatalf("Failed to create ProxyClient: %v", err)
    }
    defer client.Close()
    result := client.Connect([]Proxy {})
    if result.errors == nil {
        t.Fatalf("Expected error code %s, got nil", ERROR_NO_PROXIES_DEFINED)
    } else if result.errors[0].code != ERROR_NO_PROXIES_DEFINED {
        t.Fatalf("Expected error code %s, got %s", ERROR_NO_PROXIES_DEFINED, result.errors[0].code)
    }
    if len(result.successful) != 0 {
        t.Fatalf("Expected nil successful proxies for no proxies defined")
    }
}

func TestInvalidProxyUrl(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    defer lib.Close()
    client, err := NewProxyClient(validEthereumAddress, validStreamPartId)
    if err != nil {
        t.Fatalf("Failed to create ProxyClient: %v", err)
    }
    defer client.Close()
    proxies := []Proxy{
        Proxy{websocketUrl: invalidProxyUrl, ethereumAddress: validEthereumAddress},
    }
    fmt.Println("calling connect")
    result := client.Connect(proxies)
    fmt.Println("Testing invalid proxy URL")
    if len(result.errors) != 1 {
        t.Fatalf("Expected 1 error for invalid proxy url, got %d", len(result.errors))
    } else if result.errors[0].code != ERROR_INVALID_PROXY_URL {
        t.Fatalf("Expected error code %s, got %s", ERROR_INVALID_PROXY_URL, result.errors[0].code)
    }
    if len(result.successful) != 0 {
        t.Fatalf("Expected nil successful proxies for invalid proxy url")
    }
}

func TestInvalidProxyEthereumAddress(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    defer lib.Close()
    client, err := NewProxyClient(validEthereumAddress, validStreamPartId)
    if err != nil {
        t.Fatalf("Failed to create ProxyClient: %v", err)
    }
    defer client.Close()
    proxies := []Proxy{
        Proxy{websocketUrl: validProxyUrl, ethereumAddress: invalidEthereumAddress},
    }
    result := client.Connect(proxies)
    if len(result.errors) != 1 {
        t.Fatalf("Expected 1 error for invalid proxy ethereum address, got %d", len(result.errors))
    } else if result.errors[0].code != ERROR_INVALID_ETHEREUM_ADDRESS {
        t.Fatalf("Expected error code %s, got %s", ERROR_INVALID_ETHEREUM_ADDRESS, result.errors[0].code)
    }
    if len(result.successful) != 0 {
        t.Fatalf("Expected nil successful proxies for invalid proxy ethereum address")
    }
}

func TestProxyConnectionFailed(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    defer lib.Close()
    client, err := NewProxyClient(validEthereumAddress, validStreamPartId)
    if err != nil {
        t.Fatalf("Failed to create ProxyClient: %v", err)
    }
    defer client.Close()
    proxies := []Proxy{
        Proxy{websocketUrl: nonExistentProxyUrl0, ethereumAddress: validEthereumAddress},
    }
    result := client.Connect(proxies)
    if len(result.errors) != 1 {
        t.Fatalf("Expected 1 error for proxy connection failed, got %d", len(result.errors))
    } else if result.errors[0].code != ERROR_PROXY_CONNECTION_FAILED {
        t.Fatalf("Expected error code %s, got %s", ERROR_PROXY_CONNECTION_FAILED, result.errors[0].code)
    }
    if len(result.successful) != 0 {    
        t.Fatalf("Expected nil successful proxies for proxy connection failed")
    }
}

func TestThreeProxyConnectionsFailed(t *testing.T) {
    lib := NewLibStreamrProxyClient()
    defer lib.Close()
    client, err := NewProxyClient(validEthereumAddress, validStreamPartId)
    if err != nil {
        t.Fatalf("Failed to create ProxyClient: %v", err)
    }
    defer client.Close()
    proxies := []Proxy{
        Proxy{websocketUrl: nonExistentProxyUrl0, ethereumAddress: validEthereumAddress},
        Proxy{websocketUrl: nonExistentProxyUrl1, ethereumAddress: validEthereumAddress2},
        Proxy{websocketUrl: nonExistentProxyUrl2, ethereumAddress: validEthereumAddress3},
    }
    result := client.Connect(proxies)
    if len(result.errors) != 3 {
        t.Fatalf("Expected 3 errors for three proxy connections failed, got %d", len(result.errors))
    } else {
        for i, err := range result.errors {
            if err.code != ERROR_PROXY_CONNECTION_FAILED {
                t.Fatalf("Expected error code %s for error %d, got %s", ERROR_PROXY_CONNECTION_FAILED, i, err.code)
            }
            expectedWebsocketUrl := ""
            expectedEthereumAddress := ""
            switch i {
            case 0:
                expectedWebsocketUrl = nonExistentProxyUrl0
                expectedEthereumAddress = validEthereumAddress
            case 1:
                expectedWebsocketUrl = nonExistentProxyUrl1
                expectedEthereumAddress = validEthereumAddress2
            case 2:
                expectedWebsocketUrl = nonExistentProxyUrl2
                expectedEthereumAddress = validEthereumAddress3
            }
            if err.proxy.websocketUrl != expectedWebsocketUrl {
                t.Fatalf("Expected websocket URL %s for error %d, got %s", expectedWebsocketUrl, i, err.proxy.websocketUrl)
            }
            if err.proxy.ethereumAddress != expectedEthereumAddress {
                t.Fatalf("Expected ethereum address %s for error %d, got %s", expectedEthereumAddress, i, err.proxy.ethereumAddress)
            }
        }
    }
    if len(result.successful) != 0 {
        t.Fatalf("Expected nil successful proxies for three proxy connections failed")
    }
}